### PR TITLE
[Merged by Bors] - feat: Add API for `AlgEquiv`.

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Algebra/Equiv.lean [hunk 1]
```diff
@@ -272,6 +272,18 @@ theorem coe_algHom_injective : Function.Injective ((↑) : (A₁ ≃ₐ[R] A₂)
   fun _ _ h => ext <| AlgHom.congr_fun h
 #align alg_equiv.coe_alg_hom_injective AlgEquiv.coe_algHom_injective
 
+theorem coe_MulEquiv_injective :
+    Function.Injective ((↑) : (A₁ ≃ₐ[R] A₂) → A₁ ≃* A₂) :=
+  fun _ _ e ↦ ext fun a ↦ FunLike.congr_fun e a
```
Comment (eric-wieser) [inline Mathlib/Algebra/Algebra/Equiv.lean:277]: Do we already have the `RingEquiv`, `AddEquiv`, `Equiv`, and `LinearEquiv` versions?

File: Mathlib/Algebra/Group/Equiv/Basic.lean [hunk 2]
```diff
@@ -192,6 +192,10 @@ def MulEquivClass.toMulEquiv [Mul α] [Mul β] [MulEquivClass F α β] (f : F) :
 instance [Mul α] [Mul β] [MulEquivClass F α β] : CoeTC F (α ≃* β) :=
   ⟨MulEquivClass.toMulEquiv⟩
 
+theorem MulEquivClass.coe_MulEquiv_injective [Mul α] [Mul β] [MulEquivClass F α β] :
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Equiv/Basic.lean:195]:
```suggestion
@[to_additive]
theorem MulEquivClass.coe_MulEquiv_injective [Mul α] [Mul β] [MulEquivClass F α β] :
```

Comment (eric-wieser) [reply Mathlib/Algebra/Group/Equiv/Basic.lean:195]: Arguably it should be `toMulEquiv` not `coe_MulEquiv`

File: Mathlib/Algebra/Algebra/Equiv.lean [hunk 3]
```diff
@@ -778,6 +786,35 @@ theorem algebraMap_eq_apply (e : A₁ ≃ₐ[R] A₂) {y : R} {x : A₁} :
...
+  map_mul' := fun _ _ ↦ rfl
+
+lemma pow_toLinearMap (σ : A₁ ≃ₐ[R] A₁) (n : ℕ) :
+    (σ ^ n).toLinearMap = σ.toLinearMap ^ n :=
+  (AlgEquiv.toLinearMapHom R A₁).map_pow σ n
+
+lemma one_toLinearMap :
+    (1 : A₁ ≃ₐ[R] A₁).toLinearMap = 1 := rfl
+
+/-- The units group of `S →ₐ[R] S` is `S ≃ₐ[R] S` -/
```
Comment (eric-wieser) [inline Mathlib/Algebra/Algebra/Equiv.lean:804]: Can you add a cross-reference to the `LinearEquiv` version we already have?

Comment (erdOne) [reply Mathlib/Algebra/Algebra/Equiv.lean:804]: Do you mean `LinearMap.GeneralLinearGroup.generalLinearEquiv`?

Comment (erdOne) [reply Mathlib/Algebra/Algebra/Equiv.lean]: I'm not sure where such a lemma should go though, and I don't think this lemma is useful enough to warrant a new file of its own.

Comment (eric-wieser) [reply Mathlib/Algebra/Algebra/Equiv.lean]: Ah, I hadn't realized there was an import issue. That sounds fair.

**Comment by eric-wieser** [comment]
bors d+

Thanks!

File: Mathlib/Algebra/Algebra/Equiv.lean [hunk 4]
```diff
@@ -778,6 +782,36 @@ theorem algebraMap_eq_apply (e : A₁ ≃ₐ[R] A₂) {y : R} {x : A₁} :
...
+@[simps]
+def toLinearMapHom (R A) [CommSemiring R] [Semiring A] [Algebra R A] :
+    (A ≃ₐ[R] A) →* A →ₗ[R] A where
+  toFun := AlgEquiv.toLinearMap
+  map_one' := rfl
+  map_mul' := fun _ _ ↦ rfl
+
+lemma pow_toLinearMap (σ : A₁ ≃ₐ[R] A₁) (n : ℕ) :
+    (σ ^ n).toLinearMap = σ.toLinearMap ^ n :=
+  (AlgEquiv.toLinearMapHom R A₁).map_pow σ n
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/Algebra/Algebra/Equiv.lean:795]: There is no zpow on linear maps.

**Comment by erdOne** [comment]
bors merge

**Pull request closed without merging**